cedar-policy-validator/src/entity_manifest.rs

```diff
@@ -0,0 +1,1103 @@
+/// while [`get_expr_path`] handles the <datapath-expr>s.
+fn add_to_root_trie(
+    root_trie: &mut RootAccessTrie,
+    expr: &Expr<Option<Type>>,
+    policy_id: &PolicyID,
+    should_load_all: bool,
+) -> Result<(), EntityManifestError> {
+    match expr.expr_kind() {
+        // Literals, variables, and unkonwns without any GetAttr operations
+        // on them are okay, since no fields need to be loaded.
+        ExprKind::Lit(_) => (),
+        ExprKind::Var(_) => (),
+        ExprKind::Slot(_) => (),
+        ExprKind::Unknown(_) => return Err(EntityManifestError::PartialExpressionError),
+        ExprKind::If {
+            test_expr,
+            then_expr,
+            else_expr,
+        } => {
+            add_to_root_trie(root_trie, test_expr, policy_id, should_load_all)?;
+            add_to_root_trie(root_trie, then_expr, policy_id, should_load_all)?;
+            add_to_root_trie(root_trie, else_expr, policy_id, should_load_all)?;
+        }
+        ExprKind::And { left, right } => {
+            add_to_root_trie(root_trie, left, policy_id, should_load_all)?;
+            add_to_root_trie(root_trie, right, policy_id, should_load_all)?;
+        }
+        ExprKind::Or { left, right } => {
+            add_to_root_trie(root_trie, left, policy_id, should_load_all)?;
+            add_to_root_trie(root_trie, right, policy_id, should_load_all)?;
+        }
+        ExprKind::UnaryApp { op, arg } => match op {
+            UnaryOp::Not => add_to_root_trie(root_trie, arg, policy_id, should_load_all)?,
+            UnaryOp::Neg => add_to_root_trie(root_trie, arg, policy_id, should_load_all)?,
+        },
+        ExprKind::BinaryApp { op, arg1, arg2 } => match op {
+            // Special case! Equality between records requires
+            // that we load all fields.
+            // This could be made more precise if we check the type.
+            BinaryOp::Eq => {
+                add_to_root_trie(root_trie, arg1, policy_id, true)?;
+                add_to_root_trie(root_trie, arg1, policy_id, true)?;
+            }
+            BinaryOp::In => {
+                // Recur normally on the rhs
+                add_to_root_trie(root_trie, arg2, policy_id, should_load_all)?;
+
+                // The lhs should be a datapath expression.
+                let mut flat_slice = get_expr_path(arg1, policy_id)?;
+                flat_slice.parents_required = true;
+                *root_trie = root_trie.union(&flat_slice.to_primary_slice());
+            }
+            BinaryOp::Contains | BinaryOp::ContainsAll | BinaryOp::ContainsAny => {
+                // Like equality, another special case for records.
+                add_to_root_trie(root_trie, arg1, policy_id, true)?;
+                add_to_root_trie(root_trie, arg2, policy_id, true)?;
+            }
+            BinaryOp::Less | BinaryOp::LessEq | BinaryOp::Add | BinaryOp::Sub | BinaryOp::Mul => {
+                // These operators work on values, so no special
```

Comment:
nit: in the terminology used elsewhere in the Cedar implementation, `Value` includes records.  Might be better to say "literals" or just "integers"